Framework/Core/src/WorkflowHelpers.cxx

```diff
@@ -703,34 +703,42 @@ std::tuple<std::vector<InputSpec>, std::vector<unsigned char>> WorkflowHelpers::
     unsigned char outputtype = 0;
 
     // is AOD?
-    if (DataSpecUtils::partialMatch(outputSpec, header::DataOrigin("AOD")))
+    if (DataSpecUtils::partialMatch(outputSpec, header::DataOrigin("AOD"))) {
       outputtype += 2;
+    }
 
     // is dangling output?
     bool matched = false;
     for (size_t ii = 0, ie = inputs.size(); ii != ie; ++ii) {
       auto& input = inputs[ii];
       // Inputs of the same workflow cannot match outputs
-      if (output.workflowId == input.workflowId)
+      if (output.workflowId == input.workflowId) {
         continue;
+      }
       auto& inputSpec = workflow[input.workflowId].inputs[input.id];
       if (DataSpecUtils::match(inputSpec, outputSpec)) {
         matched = true;
         break;
       }
     }
-    if (!matched)
+    if (!matched) {
       outputtype += 1;
+    }
 
     // update results and outputtypes
     auto input = DataSpecUtils::matchingInput(outputSpec);
     char buf[64];
     input.binding = (snprintf(buf, 63, "output_%zu_%zu", output.workflowId, output.id), buf);
-    results.emplace_back(input);
 
-    outputtypes.emplace_back(outputtype);
+    // make sure that entries are unique
+    if (std::find(results.begin(), results.end(), input) == results.end()) {
+      results.emplace_back(input);
+      outputtypes.emplace_back(outputtype);
+    }
```

Comment:
This is n*log(n), no? Good enough for now, but maybe you can make it faster with an unordered_map or sort + unique.